src/arizona_template_scanner.erl

```diff
@@ -0,0 +1,548 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright 2024 Arizona Framework <contact@arizonaframe.work>
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(arizona_template_scanner).
+-moduledoc false.
+
+%% API functions.
+-export([scan/1]). -ignore_xref([scan/1]).
+
+%% Types
+-export_type([token/0]).
+-export_type([location/0]).
+-export_type([line/0]).
+-export_type([column/0]).
+-export_type([result/0]).
+-export_type([error_reason/0]).
+
+-type state() :: #{
+    line => line(),
+    column => column(),
+    first_column => non_neg_integer(),
+    position => non_neg_integer()
```

Comment:
What's position, if you have line and column?
And first_column?

Reply:
`position` is the "buffer position". It is used to get the binary part, e.g.:
```erlang
Pos = maps:get(position, State),
Expr = binary_part(Bin, Pos, Len), % Pos is position and Len is lenght
```
We are "walking" through the binary, so I need the position and length to get chunks.
Hmm `first_column` is not necessary here. I'll remove it.

Reply:
`first_column` removed in [3eec6a1](https://github.com/arizona-framework/arizona/pull/143/commits/3eec6a12c2916d05167c46e08134f0644933f3c9)